lib/flight_scheduler/schedulers/fifo_scheduler.rb

```diff
@@ -124,48 +44,89 @@ def allocate_jobs
-          enum.next
-          next
+        if candidate.job_type == 'ARRAY_JOB'
+          array_job = candidate
+          candidate = array_job.task_generator.next_task
+          if candidate.nil?
+            # The array job does not have any pending tasks left.  We
+            # shouldn't ever get here, but let's handle the case anyway.
+            next
+          end
         end
 
-        # Create the allocation
-        allocation = allocate_job(next_job)
+        allocation = allocate_job(candidate)
         if allocation.nil?
-          @data[next_group_id][:job].reason_pending = 'Resources'
-          next_job.reason_pending = 'Resources'
+          if candidate.job_type == 'ARRAY_TASK'
+            candidate.array_job.reason_pending = 'Resources'
+          else
+            candidate.reason_pending = 'Resources'
+          end
           break
         else
+          if candidate.job_type == 'ARRAY_TASK'
+            FlightScheduler.app.job_registry.add(candidate)
+            candidate.array_job.task_generator.advance_next_task
+          end
           FlightScheduler.app.allocations.add(allocation)
           new_allocations << allocation
-          @data[next_group_id][:active] << next_job
-          enum.next
         end
       end
+
+      # We've exited the allocation loop. As this is a FIFO, any jobs left
+      # 'WaitingForScheduling' are blocked on priority.  We'll update a few of
+      # them to show that is the case.
+      #
+      # A more complicated scheduler would likely do this whilst iterating
+      # over the jobs.
+      queue[0...5]
+        .select { |job| job.pending? && job.allocation.nil? }
+        .select { |job| job.reason_pending == 'WaitingForScheduling' }
+        .each { |job| job.reason_pending = 'Priority' }
     end
     new_allocations
   end
 
-  private
+  def queue
+    # For the FIFO scheduler, the queue is sorted and slightly filtered view
+    # of the registry of jobs.
+    #
+    # The sorting and filtering is primarily intended for display purposes,
+    # but also using it for processing ensures that the jobs are processed in
+    # the order in which the queue output suggests.
+    #
+    # The jobs are sorted to ensure that running jobs are placed higher in the
+    # queue output than pending jobs.
+    #
+    # The jobs are filtered to ensure that array jobs without any pending
+    # tasks are not included.
+    #
+    # A more complicated scheduler may implement a priority queue and need to
+    # keep its own record of the queued jobs rather than using job registry.
+    running_jobs_first = lambda { |job| job.reason_pending.nil? ? -1 : 1 }
+
+    # NOTE: We rely on the sort being stable to ensure that earlier added jobs
+    # are considered prior to later added jobs.
+    FlightScheduler.app.job_registry.jobs
+      .reject { |j| j.job_type == 'ARRAY_JOB' && j.task_generator.finished? }
+      .reject { |j| %w(COMPLETED CANCELLED FAILED).include?(j.state) }
```

Comment:
The previous two lines could probably be replaced by a `finished?` flag on the `Job`

Reply:
Perhaps.  Though that raises the question "what does it mean for a job to be finshed?".  For `JOB` and `ARRAY_TASK` that is very clear; its in one of its terminal states.

It's not so clear for an `ARRAY_JOB`.  Does it mean that the task generator has finished?  Or does it mean that the job is in one of its terminal states?  Or does it mean both of them must be true?  Or does it mean that either of them must be true?  All four possibilities are possible.

I suspect that whatever one we would select would prove to be the wrong option and a confusing option as often as it proved to be helpful.